README.md
# TempGopher

TempGopher is a thermostat application written in Go. It turns a Raspberry Pi into a thermostat you can control with your web browser.

## Requirements

You will need a Raspberry Pi with the following components:

* A network connection
* DS18B120, 1-wire temperature sensors
* GPIO pins
* Relays for powering on/off your equipment hooked up to the GPIO pins

## Install

You will need to setup your Raspberry Pi for this to work.

1. Setup your DS18B20 temperature sensor. [Adafruit has a good tutorial for getting them working](https://learn.adafruit.com/adafruits-raspberry-pi-lesson-11-ds18b20-temperature-sensing/hardware)
2. Connect your relay switches to the GPIO pins
3. [Download the install.sh script to your Raspberry Pi](https://gitlab.com/shouptech/tempgopher/-/jobs/artifacts/master/raw/install.sh?job=build)
4. Run the script! The script will download the latest binary and configure the thermostat with some initial values.
5. After configuration, point your web browser to the configured URL.

## Configuration Script

You will be asked some questions during the initial configuration of TempGopher. You also see some defaults in brackets. If the brackets look good, just hit enter.

* `Listen address?` - The address & port TempGopher should listen on. Omitting the address and just specifying the port means it listens on all addresses. Default is `:8080`.
* `Base URL?` - This is what you will type in to your web browser to access TempGopher. If you don't have DNS configured, should probably be `http://<ipaddress>:8080`
* `Display temperature in fahrenheit?` - Set to true if you want fahrenheit, otherwise defaults to celsius.
* `Configure sensor w/ ID: 28-xxxxx` - If you set up your DS18B20 sensors correctly, you should see it's ID listed. Enter `Y` and answer the prompts to configure it. If you have multiple sensors, you will be asked this question multiple times.
* `Sensor alias:` - Name to display in the web browser for this sensor.
* `High temperature:` - The high temperature to kick the cooling on.
* `Cooling minutes:` - The number of minutes to run the cooler once the temperature is below the High temperature threshold.
* `Cooling GPIO:` - The pin your cooling relay switch is hooked into.
* `Invert cooling switch` - If set to `true`, the cooling will be ON when the switch is LOW. This should usually be `false`, so that is the default
* `Low temperature:` - The low temperature to kick the heating on.
* `Heating minutes:` - The number of minutes to run the heater once the temperature is below the Low temperature threshold.
* `Heating GPIO:` - The pin your heating relay switch is hooked into.
* `Invert heating switch` - If set to `true`, the heating will be ON when the switch is LOW. This should usually be `false`, so that is the default
* `Enable verbose logging` - If set to `true`, TempGopher will display in the console every thermostat reading. This can be quite verbose, so the default is `false`.
* `Write data to an Influx database?` - Whether or not to configure an Influx database
* `Enable user authentication?` - Whether or not to enable authentication

## Example configuration script

```
$ bash install.sh

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   147  100   147    0     0    357      0 --:--:-- --:--:-- --:--:--   356

TempGopher v0.2.0
You will now be asked a series of questions to help configure your thermostat.
Don't worry, it will all be over quickly.

Default values will be in brackets. Just press enter if they look good.
=====
Listen address?
[:8080]:
Base URL? (This is what you type into your browser to get to the web UI)
[http://beerpi:8080]: http://10.30.14.130:8080
Display temperatures in fahrenheit? (Otherwise uses celsius)
[true]:
Configure sensor w/ ID: 28-000008083108
[Y/n]:
Sensor alias: fermenter
High temperature: 20
Cooling minutes: 4
Cooling GPIO: 19
Invert cooling switch [false]:
Low temperature: 19
Heating minutes: 0.5
Heating GPIO: 13
Invert heating switch [false]:
Enable verbose logging [false]:
Write data to an Influx database?
[Y/n]: y
Influx address [http://influx:8086]:
Influx Username []:
Influx Password []:
Influx UserAgent [InfluxDBClient]:
Influx timeout (in seconds) [30]:
Influx database []: tempgopher
Enable InsecureSkipVerify? [fasle]:
Username: mike
Password: ********
Add another user? [y/N]: y
Username: foo
Password: ***
Add another user? [y/N]: n

```


CHANGELOG.md
# TempGopher Changelog

## 0.4.0

Release 2018-11-01

* Enable/disable heating or cooling via UI
* Improved unit tests!
* Fixed bug where updating/clearing changes would cause multiple refreshes in a row [#17](https://gitlab.com/shouptech/tempgopher/issues/17)

## 0.3.1

Release: 2018-10-24

* Fixes bug where UI would not display two sensors correctly [#15](https://gitlab.com/shouptech/tempgopher/issues/15)
* Fixes a typo during the CLI configuration [#14](https://gitlab.com/shouptech/tempgopher/issues/14)

## 0.3.0

Release: 2018-10-20

* You can now supply a list of users for simple authentication
* Will write data to an Influx DB if configured
* Adds the ability to selectively disable heating or cooling
* Checks for the existence of a config file before generating a new one

## 0.2.0

Release: 2018-10-11

* You can now update the configuration using the UI or API
* A script to help install has been added
* The binary will now generate a usable configuration

## 0.1.1

Release: 2018-10-07

* Changes temperature logic. See #8. Fixes a situation where temperature 'floats' at the threshold and the switch is rapidly cycled.

## 0.1.0

Released: 2018-10-04

* Added single page Web UI, packaged into app
* Moved to GitLab, uses GitLab CI to run builds

## 0.0.1

Released: 2018-10-02

* Initial Release
* API only
* Supports multiple thermostats


cli.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"

	"github.com/howeyc/gopass"
	"github.com/yryz/ds18b20"
)

// ReadInput reads the next line from a Reader. It will return 'def' if nothing
// was input, or it will return the response.
func ReadInput(r *bufio.Reader, def string) string {
	resp, err := r.ReadString('\n')
	if err != nil {
		panic(err)
	}

	if resp == "\n" {
		return def
	}

	return resp[:len(resp)-1]
}

// ParsePort returns the port number from a listen address
func ParsePort(addr string) (uint16, error) {
	parts := strings.Split(addr, ":")

	port, err := strconv.ParseUint(parts[len(parts)-1], 10, 16)

	if err != nil {
		return 0, err
	}

	return uint16(port), nil
}

// PromptForConfiguration walks a user through configuration
func PromptForConfiguration(in io.Reader) Config {
	reader := bufio.NewReader(in)

	var config Config

	fmt.Printf("TempGopher v%s\n", Version)
	fmt.Println("You will now be asked a series of questions to help configure your thermostat.")
	fmt.Println("Don't worry, it will all be over quickly.")
	fmt.Println("\nDefault values will be in brackets. Just press enter if they look good.")
	fmt.Println("=====")

	fmt.Print("Listen address?\n[:8080]: ")
	config.ListenAddr = ReadInput(reader, ":8080")

	port, err := ParsePort(config.ListenAddr)
	if err != nil {
		panic(err)
	}
	hostname, err := os.Hostname()
	if err != nil {
		panic(err)
	}
	fmt.Println("Base URL? (This is what you type into your browser to get to the web UI)")
	defURL := fmt.Sprintf("http://%s:%d", hostname, port)
	fmt.Printf("[%s]: ", defURL)
	config.BaseURL = ReadInput(reader, defURL)

	fmt.Println("Display temperatures in fahrenheit? (Otherwise uses celsius)")
	fmt.Print("[true]: ")

	config.DisplayFahrenheit, err = strconv.ParseBool(ReadInput(reader, "true"))
	if err != nil {
		panic(err)
	}

	// Configure sensors
	sensors, err := ds18b20.Sensors()
	if err != nil {
		fmt.Println("Couldn't find any sensors. Did you enable the 1-wire bus?")
		fmt.Printf("The error was: %s\n", err)
		os.Exit(1)
	}

	for _, sensor := range sensors {
		fmt.Printf("Configure sensor w/ ID: %s\n", sensor)
		fmt.Print("[Y/n]: ")
		choice := ReadInput(reader, "y")
		if strings.ToLower(choice)[0] != 'y' {
			continue
		}

		var s Sensor
		s.ID = sensor

		fmt.Print("Sensor alias: ")
		s.Alias = ReadInput(reader, "")
		if s.Alias == "" {
			panic("Alias cannot be blank")
		}

		fmt.Print("Disable cooling? [false]: ")
		s.CoolDisable, err = strconv.ParseBool(ReadInput(reader, "false"))
		if err != nil {
			panic(err)
		}

		if !s.CoolDisable {

			fmt.Print("High temperature: ")
			s.HighTemp, err = strconv.ParseFloat(ReadInput(reader, ""), 64)
			if err != nil {
				panic(err)
			}

			fmt.Print("Cooling minutes: ")
			s.CoolMinutes, err = strconv.ParseFloat(ReadInput(reader, ""), 64)
			if err != nil {
				panic(err)
			}

			fmt.Print("Cooling GPIO: ")
			resp, err := strconv.ParseInt(ReadInput(reader, ""), 10, 32)
			s.CoolGPIO = int32(resp)
			if err != nil {
				panic(err)
			}

			fmt.Print("Invert cooling switch [false]: ")
			s.CoolInvert, err = strconv.ParseBool(ReadInput(reader, "false"))
			if err != nil {
				panic(err)
			}
		}

		fmt.Print("Disable heating? [false]: ")
		s.HeatDisable, err = strconv.ParseBool(ReadInput(reader, "false"))
		if err != nil {
			panic(err)
		}

		if !s.HeatDisable {

			fmt.Print("Low temperature: ")
			s.LowTemp, err = strconv.ParseFloat(ReadInput(reader, ""), 64)
			if err != nil {
				panic(err)
			}

			fmt.Print("Heating minutes: ")
			s.HeatMinutes, err = strconv.ParseFloat(ReadInput(reader, ""), 64)
			if err != nil {
				panic(err)
			}

			fmt.Print("Heating GPIO: ")
			resp, err := strconv.ParseInt(ReadInput(reader, ""), 10, 32)
			s.HeatGPIO = int32(resp)
			if err != nil {
				panic(err)
			}

			fmt.Print("Invert heating switch [false]: ")
			s.HeatInvert, err = strconv.ParseBool(ReadInput(reader, "false"))
			if err != nil {
				panic(err)
			}
		}

		fmt.Print("Enable verbose logging [false]: ")
		s.Verbose, err = strconv.ParseBool(ReadInput(reader, "false"))
		if err != nil {
			panic(err)
		}

		config.Sensors = append(config.Sensors, s)
	}

	fmt.Println("Write data to an Influx database?")
	fmt.Print("[Y/n]: ")
	choice := ReadInput(reader, "y")
	if strings.ToLower(choice)[0] == 'y' {
		fmt.Print("Influx address [http://influx:8086]: ")
		config.Influx.Addr = ReadInput(reader, "http://influx:8086")

		fmt.Print("Influx Username []: ")
		config.Influx.Username = ReadInput(reader, "")

		fmt.Print("Influx Password []: ")
		config.Influx.Password = ReadInput(reader, "")

		fmt.Print("Influx UserAgent [InfluxDBClient]: ")
		config.Influx.UserAgent = ReadInput(reader, "InfluxDBClient")

		fmt.Print("Influx timeout (in seconds) [30]: ")
		config.Influx.Timeout, err = strconv.ParseFloat(ReadInput(reader, "30"), 64)
		if err != nil {
			panic(err)
		}

		fmt.Print("Influx database []: ")
		config.Influx.Database = ReadInput(reader, "")

		fmt.Print("Enable InsecureSkipVerify? [false]: ")
		config.Influx.InsecureSkipVerify, err = strconv.ParseBool(ReadInput(reader, "false"))
		if err != nil {
			panic(err)
		}
	}

	fmt.Println("Enable user authentication?")
	fmt.Print("[Y/n]: ")
	choice = ReadInput(reader, "y")
	if strings.ToLower(choice)[0] == 'y' {
		another := true
		for another {
			fmt.Print("Username: ")
			username := ReadInput(reader, "")
			fmt.Print("Password: ")
			password, err := gopass.GetPasswdMasked()
			if err != nil {
				panic(err)
			}
			config.Users = append(config.Users, User{username, string(password)})

			fmt.Print("Add another user? [y/N]: ")
			choice = ReadInput(reader, "n")
			if strings.ToLower(choice)[0] == 'n' {
				another = false
			}
		}
	}

	return config
}

// ConfigCLI prompts the user for configuration and writes to a config file
func ConfigCLI(path string) {
	// Check if path exists
	if _, err := os.Stat(path); !os.IsNotExist(err) {
		fmt.Printf("File exists, or some other error trying to open file %s\n", path)
		os.Exit(1)
	}

	config := PromptForConfiguration(os.Stdin)

	SaveConfig(path, config)
}
